src/librawspeed/decoders/StiDecoder.cpp

```diff
@@ -0,0 +1,97 @@
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+#include "decoders/StiDecoder.h"
+#include "adt/Point.h"                              // for iPoint2D
+#include "decoders/RawDecoderException.h"           // for ThrowException
+#include "decompressors/HasselbladLJpegDecoder.h"   // for HasselbladLJpegDec...
+#include "decompressors/UncompressedDecompressor.h" // for UncompressedDeco...
+#include "io/Buffer.h"                              // for Buffer, DataBuffer
+#include "io/ByteStream.h"                          // for ByteStream
+#include "io/Endianness.h"                          // for Endianness, Endian...
+#include "metadata/ColorFilterArray.h"              // for CFAColor, CFAColor...
+#include "tiff/TiffEntry.h"                         // for TiffEntry
+#include "tiff/TiffIFD.h"                           // for TiffRootIFD, TiffIFD
+#include "tiff/TiffTag.h"                           // for TiffTag, TiffTag::...
+#include <array>                                    // for array
+#include <cstdint>                                  // for uint32_t
+#include <memory>                                   // for unique_ptr, allocator
+#include <string>                                   // for operator==, string
+
+namespace rawspeed {
+
+class CameraMetaData;
+
+bool StiDecoder::isAppropriateDecoder(const TiffRootIFD* rootIFD,
+                                      [[maybe_unused]] Buffer file) {
+  const auto id = rootIFD->getID();
+  const std::string& make = id.make;
+
+  // FIXME: magic
+
+  return make == "Sinar AG";
+}
+
+RawImage StiDecoder::decodeRawInternal() {
+  const auto* raw = mRootIFD->getIFDWithTag(TiffTag::TILEOFFSETS, 0);
+  uint32_t width = raw->getEntry(TiffTag::IMAGEWIDTH)->getU32();
+  uint32_t height = raw->getEntry(TiffTag::IMAGELENGTH)->getU32();
+  uint32_t compression = raw->getEntry(TiffTag::COMPRESSION)->getU32();
+
+  mRaw->dim = iPoint2D(width, height);
+
+  if (1 != compression)
+    ThrowRDE("Unexpected compression type.");
+
+  DecodeUncompressed(raw);
+  return mRaw;
+}
+
+void StiDecoder::DecodeUncompressed(const TiffIFD* raw) const {
+  if (mRaw->getDataType() != RawImageType::UINT16)
+    ThrowRDE("Unexpected data type");
+
+  if (mRaw->getCpp() != 1 || mRaw->getBpp() != sizeof(uint16_t))
+    ThrowRDE("Unexpected cpp: %u", mRaw->getCpp());
+
+  // FIXME: could be wrong.
+  if (!mRaw->dim.hasPositiveArea() || mRaw->dim.x % 2 != 0 ||
+      mRaw->dim.y % 2 != 0 || mRaw->dim.x > 4992 || mRaw->dim.y > 6668) {
```

Comment:
Is the RPU sample in portrait by any chance? Is it possible that the camera outputs an already oriented raw raster? The [spec](https://sinar.swiss/assets/Uploads/Manual-Sinarback-eVolution-EN.pdf) says 6668 x 4992 4:3 (landscape?), not the other way around...

Reply:
> Is the RPU sample in portrait by any chance?

@robert-olding ?

Reply:
Note also that the raw IFD has:
```
[IFD1]          Orientation                     : Rotate 180
```
vs. the no-op oreientation of the root IFD:
```
[IFD0]          Orientation                     : Horizontal (normal)
```

Reply:
Yes. The shot was taken with the digital back oriented in the vertical (portrait) position.

Reply:
@robert-olding Can you share (upload to RPU) a landscape sample as well please? If not (you mentioned privacy issues), can you at least share the output of e.g. `exiftool -a -u -s -g1 -w txt <landscape_file>.sti`?

Reply:
`Addy_and_Odin_Image_0095.sti` is also portrait-oriented apparently.
I'm guessing the camera always writes images that way.

Reply:
The Addy_and_Odin file were shot with the digital back oriented in the landscape position

Reply:
That's what i said, yes.

Reply:
So the sensor really is scanned that way it would seem 

However, I still need the various Orientation tags from the landscape case.

Reply:
`[IFD1]          Orientation                     : Rotate 270 CW`
[Addy_and_Odin_Image_0095.txt](https://github.com/darktable-org/rawspeed/files/13078013/Addy_and_Odin_Image_0095.txt)